### ISSUE TITLE
Warnings when bulding with cl.exe (llvm::MVT::SimpleValueType ?)

### ISSUE THREAD

**Issue by mulle-nat** [llvm/llvm-project#53018]
This warning is in German it translates to "truncation of a constant value".  (Vorlage means template). Sorry I don't have an english system for windows here. It's probably harmless, but it's annoying, because it's in many files. I just picked one out:

```
[2699/4830] Building CXX object lib\Target\RISCV\CMakeFiles\LLVMRISCVCodeGen.dir\RISCVSubtarget.cpp.obj
E:\home\src\srcL\mulle-clang-130\mulle-clang-project\llvm\include\llvm/ADT/Sequence.h(35): warning C4309: "static_cast": Verkürzung eines konstanten Wertes
E:\home\src\srcL\mulle-clang-130\mulle-clang-project\llvm\include\llvm/ADT/Sequence.h(48): note: Siehe Verweis auf die gerade kompilierte Instanziierung "bool llvm::detail::canTypeFitValue<intmax_t,Integral>(const U)" der Funktions-Vorlage.
        with
        [
            Integral=type,
            U=type
        ]
E:\home\src\srcL\mulle-clang-130\mulle-clang-project\llvm\include\llvm/ADT/Sequence.h(60): note: Siehe Verweis auf die gerade kompilierte Instanziierung "llvm::detail::CheckedInt llvm::detail::CheckedInt::from<type,false>(Integral)" der Funktions-Vorlage.
        with
        [
            Integral=type
        ]
E:\home\src\srcL\mulle-clang-130\mulle-clang-project\llvm\include\llvm/ADT/Sequence.h(113): note: Siehe Verweis auf die gerade kompilierte Instanziierung "llvm::detail::CheckedInt llvm::detail::CheckedInt::from<T,false>(Enum)" der Funktions-Vorlage.
        with
        [
            T=llvm::MVT::SimpleValueType,
            Enum=llvm::MVT::SimpleValueType
        ]
E:\home\src\srcL\mulle-clang-130\mulle-clang-project\llvm\include\llvm/ADT/Sequence.h(113): note: beim Kompilieren der Klasse Vorlage-Memberfunktion "llvm::detail::SafeIntIterator<llvm::MVT::SimpleValueType,false>::SafeIntIterator(T)"
        with
        [
            T=llvm::MVT::SimpleValueType
        ]
E:\home\src\srcL\mulle-clang-130\mulle-clang-project\llvm\include\llvm/ADT/Sequence.h(191): note: Siehe Verweis auf die gerade kompilierte Instanziierung "llvm::detail::SafeIntIterator<llvm::MVT::SimpleValueType,false>::SafeIntIterator(T)" der Funktions-Vorlage.
        with
        [
            T=llvm::MVT::SimpleValueType
        ]
E:\home\src\srcL\mulle-clang-130\mulle-clang-project\llvm\include\llvm/ADT/Sequence.h(212): note: Siehe Verweis auf die gerade kompilierte Klasse Vorlage-Instanziierung "llvm::detail::SafeIntIterator<llvm::MVT::SimpleValueType,false>".
E:\home\src\srcL\mulle-clang-130\mulle-clang-project\llvm\include\llvm/ADT/Sequence.h(234): note: Siehe Verweis auf die gerade kompilierte Klasse Vorlage-Instanziierung "llvm::iota_range<T>".
        with
        [
            T=llvm::MVT::SimpleValueType
        ]
E:\home\src\srcL\mulle-clang-130\mulle-clang-project\llvm\include\llvm/Support/MachineValueType.h(1408): note: Siehe Verweis auf die gerade kompilierte Instanziierung "llvm::iota_range<T> llvm::seq_inclusive<llvm::MVT::SimpleValueType>(T,T)" der Funktions-Vorlage.
        with
        [
            T=llvm::MVT::SimpleValueType
        ]
```

This is with `Microsoft (R) C/C++-Optimierungscompiler Version 19.23.28106.4 für x86
Copyright (C) Microsoft Corporation. Alle Rechte vorbehalten.`:

**Comment by jrtc27** [comment]
I doubt it's specific to RISC-V given the "because it's in many files. I just picked one out" in the bug report

**Comment by jrtc27** [comment]
It's complaining that:
```
template <typename T, typename U> bool canTypeFitValue(const U Value) {
  const intmax_t BotT = intmax_t(std::numeric_limits<T>::min());
  const intmax_t BotU = intmax_t(std::numeric_limits<U>::min());
  const uintmax_t TopT = uintmax_t(std::numeric_limits<T>::max());
  const uintmax_t TopU = uintmax_t(std::numeric_limits<U>::max());
  return !((BotT > BotU && Value < static_cast<U>(BotT)) ||
           (TopT < TopU && Value > static_cast<U>(TopT)));
}
```
C4309 in English is truncation of a constant value, which is entirely expected here, surely.

So not a RISC-V backend issue, just an ADT issue.

**Comment by jrtc27** [comment]
Note that this is a false positive: if the first cast truncates then BotU > BotT by definition so the first && is false, and if the second cast truncates then TopU < TopT by definition so the second && is false. Both are short-circuiting.
